pkg/validate/validate_test.go

```diff
@@ -324,3 +323,114 @@ func TestValidate(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateUpdateContainerService(t *testing.T) {
+
+	var twoAgents = []byte(`---
+location: eastus
+name: openshift
+properties:
+  routerProfiles:
+  - name: default
+    publicSubdomain: test.example.com
+  agentPoolProfiles:
+  - name: infra
+    role: infra
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  - name: myCompute
+    role: compute
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+`)
+	var threeAgents = []byte(`---
+location: eastus
+name: openshift
+properties:
+  routerProfiles:
+  - name: default
+    publicSubdomain: test.example.com
+  agentPoolProfiles:
+  - name: infra
+    role: infra
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  - name: myCompute
+    role: compute
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  - name: mySomethingElse
+    role: compute
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+`)
+
+	tests := map[string]struct {
+		newTemplate   []byte
+		newAgentCount int
+		oldTemplate   []byte
+		oldAgentCount int
+		externalOnly  bool
+		wantErrs      []error
+	}{
+		"good-2": {
+			newTemplate:   twoAgents,
+			newAgentCount: 2,
+			oldTemplate:   twoAgents,
+			oldAgentCount: 1,
+			externalOnly:  false, // not used
+			wantErrs:      nil,
+		},
+		"good-3": {
+			newTemplate:   threeAgents,
+			newAgentCount: 2,
+			oldTemplate:   threeAgents,
+			oldAgentCount: 1,
+			externalOnly:  false, // not used
+			wantErrs:      nil,
+		},
+		"bad-2-3": {
+			newTemplate:   twoAgents,
+			newAgentCount: 2,
+			oldTemplate:   threeAgents,
+			oldAgentCount: 1,
+			externalOnly:  false, // not used
+			wantErrs:      []error{errors.New("invalid change [Properties.AgentPoolProfiles.slice[2]: <no value> != {mySomethingElse 1 Standard_D2s_v3  Linux compute}]")},
+		},
+	}
+
+	for name, tt := range tests {
+		var newOc *v20180930preview.OpenShiftManagedCluster
+		err := yaml.Unmarshal(tt.newTemplate, &newOc)
+		if err != nil {
+			t.Fatal(err)
+		}
+		newCs := api.ConvertFromV20180930preview(newOc)
+		for _, o := range newCs.Properties.AgentPoolProfiles {
+			o.Count = tt.newAgentCount
```

Comment:
unfortunately this code has no side effect (imo this is the biggest gotcha in Golang).
see https://bryce.is/writing/code/jekyll/update/2015/11/01/3-go-gotchas.html

Reply:
thanks for the link, interesting..